A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_FreeText.js

```diff
@@ -0,0 +1,138 @@
+                [""],
+                [`${store["msg_line4"] != "" ? store["msg_line4"] : "["}[color]blue`, `${store["msg_line4"] != "" ? "" : "]"}[color]blue`],
+                ["RETURN TO", `${store["sendStatus"]}`],
+                ["<AOC MENU", "SEND*[color]blue"]
+            ]);
+        }
+        updateView();
+
+        mcdu.onLeftInput[0] = () => {
+            let value = mcdu.inOut;
+            mcdu.clearUserInput();
+            if (value == "CLR") {
+                store["msg_to"] = "";
+            } else {
+                store["msg_to"] = value;
+            }            
+            CDUAocFreeText.ShowPage(mcdu, store);
+        }
+
+        mcdu.onLeftInput[1] = () => {
+            let value = mcdu.inOut;
+            mcdu.clearUserInput();
+            if (value == "CLR") {
+                store["msg_line1"] = "";
+            } else {
+                store["msg_line1"] = value;
+            }            
+            CDUAocFreeText.ShowPage(mcdu, store);
+        }
+
+        mcdu.onLeftInput[2] = () => {
+            let value = mcdu.inOut;
+            mcdu.clearUserInput();
+            if (value == "CLR") {
+                store["msg_line2"] = "";
+            } else {
+                store["msg_line2"] = value;
+            }
+            CDUAocFreeText.ShowPage(mcdu, store);
+        }
+
+        mcdu.onLeftInput[3] = () => {
+            let value = mcdu.inOut;
+            mcdu.clearUserInput();
+            if (value == "CLR") {
+                store["msg_line3"] = "";
+            } else {
+                store["msg_line3"] = value;
+            }     
+            CDUAocFreeText.ShowPage(mcdu, store);
+        }
+
+        mcdu.onLeftInput[4] = () => {
+            let value = mcdu.inOut;
+            mcdu.clearUserInput();
+            if (value == "CLR") {
+                store["msg_line4"] = "";
+            } else {
+                store["msg_line4"] = value;
+            }     
+            CDUAocFreeText.ShowPage(mcdu, store);
+        }
+
+        mcdu.onRightInput[5] = async () => {
+            const telexID = SimVar.GetSimVarValue("L:A32NX_Telex_ID", "Number");
+            const telexFlight = SimVar.GetSimVarValue("ATC FLIGHT NUMBER", "string", "FMC");
+
+            console.log(telexFlight);
+            if (telexID != "-1" && telexFlight != "") {
+                store["sendStatus"] = "QUEUED"
+                updateView();
+                const recipient = store["msg_to"]
+                const msgLines = [store["msg_line1"], store["msg_line2"], store["msg_line3"], store["msg_line4"]].join(";");
+                mcdu.clearUserInput();
+                let errors = 0;
+                let endpoint = "https://api.flybywiresim.com/txmsg";
+
+                const getData = async () => {
+                    if (recipient != "" && msgLines != ";;;") {
+                        await fetch(`${endpoint}?from=${telexFlight}&to=${recipient}&message=${msgLines}`, {method: "POST"})
```

Comment:
I hope there is sanitisation of user input at the server end, i.e. preventing special characters, maximum message length, etc. else people will spam arabic unicode and webdings copypastas. Don't think this is open to XSS.

Reply:
Yep, I'll implement a server-side regex filter tomorrow to only allow alphanumeric characters, periods, and forward slashes, as well as a filter to replace slurs with asterisks.